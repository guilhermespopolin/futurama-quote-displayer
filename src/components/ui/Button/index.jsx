import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const StyledButton = styled.button`
  display: inline-flex;
  align-items: center;
  padding: 0.5rem;
  font-family: 'Oswald', sans-serif;
  font-size: 1.1rem;
  color: ${({ theme }) => theme.colors.white};
  background-color: ${({ theme }) => theme.colors.primary0};
  border: none;
  outline: none;
  border-radius: 3px;
  min-width: 60px;

  &:hover:enabled {
    cursor: pointer;
    background-color: ${({ theme }) => theme.colors.primary1};
  }

  &:disabled {
    opacity: 0.4;
  }

  &:disabled:hover {
    cursor: not-allowed;
  }
`

function Button({
  onClick,
  children,
  icon,
  iconRight,
  ...others
}) {
  const content = iconRight ? (
    <Fragment>
      {children}
      {icon ? React.cloneElement(icon, { style: { marginLeft: '0.5rem' } }) : null}
    </Fragment>
  ) : (
    <Fragment>
      {icon ? React.cloneElement(icon, { style: { marginRight: '0.5rem' } }) : null}
      {children}
    </Fragment>
  )

  return (
    <StyledButton onClick={onClick} {...others}>
      {content}
    </StyledButton>
  )
}

Button.propTypes = {
  onClick: PropTypes.func,
  children: PropTypes.node,
  icon: PropTypes.element,
  iconRight: PropTypes.bool,
}

Button.defaultProps = {
  onClick: () => {},
  children: 'Click me!',
  icon: null,
  iconRight: false,
}

export default Button
