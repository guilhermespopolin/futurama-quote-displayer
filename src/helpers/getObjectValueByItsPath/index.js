/**
 *  Returns object value given its path.
 *
 * @param {object} srcObject - Source object
 * @param {string} path - value path
 * @param {any} defaultValue - optional. Default value to be returned
 * in case value `path` doesnt exist on provided `srcObject`.
 */
function getObjectValueByItsPath(srcObject, path, defaultValue) {
  const dottedSplittedPath = path.split('.')
  const dottedSplittedPathLength = dottedSplittedPath.length
  const key = dottedSplittedPathLength === 1 ? path : dottedSplittedPath[0]

  if (!Object.prototype.hasOwnProperty.call(srcObject, key)) {
    return defaultValue
  }

  if (dottedSplittedPathLength === 1) {
    return srcObject[key]
  }

  const nextPath = dottedSplittedPath.slice(1).join('.')

  return getObjectValueByItsPath(srcObject[key], nextPath, defaultValue)
}

export default getObjectValueByItsPath
