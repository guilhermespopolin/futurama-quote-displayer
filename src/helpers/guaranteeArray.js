export default function guarateeArray(supostArray) {
  return Array.isArray(supostArray) ? supostArray : [supostArray]
}
