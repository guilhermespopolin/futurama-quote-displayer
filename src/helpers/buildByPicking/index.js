import getObjectValueByItsPath from 'helpers/getObjectValueByItsPath'

/**
 * Given a reference object, built a new one by taking some of its properties.
 *
 * @param {object} srcObject - Object that should used as reference.
 * @param {{ key: string, path: string }[]} pickMapList - List of objects that describes
 * which properties should be taken from `srcObject` (path) and under which property of on
 * construction object its value should be placed on (key).
 *
 * If provided, `path` should be on dotted format (e.g addres.city). In case it's not
 * provided, `key` will be used as location.
 *
 * Examples:
 *
 * const person = { name: 'John Doe', address: { city: 'Goron', zicode: '11111' } }
 *
 * buildByPicking(person, [{ key: 'name' }]) --> { name: 'John Doe' }
 * buildByPicking(person, [{ key: 'fullname', path: 'name' }]) --> { fullname: 'John Doe' }
 * buildByPikcing(person, [{ key: 'name' }, { key: 'city', path: 'address.city' }])
 *  --> { name: 'John Doe', city: 'Goron' }
 */
const buildByPicking = (srcObject, pickMapList) => pickMapList.reduce((builtObject, pickMap) => ({
  ...builtObject,
  [pickMap.key]: getObjectValueByItsPath(srcObject, pickMap.path || pickMap.key),
}), {})

export default buildByPicking
