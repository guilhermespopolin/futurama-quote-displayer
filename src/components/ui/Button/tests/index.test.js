import React from 'react'
import { render } from 'helpers/tests'

import Button from '..'

describe('<Button />', () => {
  it('should render properly', () => {
    const { getByText } = render(<Button>Click Me</Button>)
    const buttonNode = getByText('Click Me')

    expect(buttonNode).toMatchSnapshot()
  })
})

