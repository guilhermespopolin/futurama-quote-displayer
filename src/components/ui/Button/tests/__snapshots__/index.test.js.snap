// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render properly 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.5rem;
  color: #fefefe;
  background-color: #00afc8;
  border: none;
  outline: none;
  border-radius: 3px;
  min-width: 60px;
}

.c0:hover:enabled {
  cursor: pointer;
  background-color: #10defb;
}

.c0:disabled {
  opacity: 0.4;
}

.c0:disabled:hover {
  cursor: not-allowed;
}

<button
  class="c0"
>
  Click Me
</button>
`;
